### ISSUE TITLE
SetSystemProperty / SetEnvironmentVariable broken for nested classes

### ISSUE THREAD

**Issue by fknittel** [junit-pioneer/junit-pioneer#480]
Looks like 3ecf1338f552d99a5192999ee142d09760413d78 had the side effect of breaking SetSystemProperty and SetEnvironmentVariable when used in nested classes. Example for SetSystemProperty:

```java
@SetSystemProperty(key = "PROP", value = "PROP-VALUE")
class NestedSetSystemPropertyTest {
    @Nested
    @TestMethodOrder(MethodOrderer.OrderAnnotation.class)
    class nested {
        @Test
        @Order(1)
        void worksAsExpected() {
            assertEquals("PROP-VALUE", System.getProperty("PROP"));
        }

        @Test
        @Order(2)
        void breaksUnexpectedly() {
            assertEquals("PROP-VALUE", System.getProperty("PROP"));
        }
    }
}
```

This is due to `SystemPropertyExtension.entriesToSet()` now only returning the property "PROP" for the outermost `ExtensionContext` in `AbstractEntryBasedExtension.beforeEach()`. While this is fine during the setup phase (beforeEach/beforeAll) during tear-down `AbstractEntryBasedExtension.restoreOriginalEntries()` uses `ExtensionContext.Store.get()` to return the matching `EntriesBackup`: https://github.com/junit-pioneer/junit-pioneer/blob/3ecf1338f552d99a5192999ee142d09760413d78/src/main/java/org/junitpioneer/jupiter/AbstractEntryBasedExtension.java#L149-L151 

Note that `ExtensionContext.Store.get()` has the following [documented behaviour](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/ExtensionContext.Store.html#get(java.lang.Object,java.lang.Class)):
> If no value is stored in the current ExtensionContext for the supplied key, ancestors of the context will be queried for a value with the same key in the Namespace used to create this store.

So although the inner class and inner methods have no `EntriesBackup`, `AbstractEntryBasedExtension.restoreOriginalEntries()` grabs the outer context's instance during tear-down of `worksAsExpected()`, which accidentally clears the properties for the `breaksUnexpectedly()` test.

It would be great if `ExtensionContext.Store` were to provide a way to only query the store of the provided context and not the ancestor stores too. Haven't come up with an elegant solution for this yet. A workaround could be to store something that identifies the originating context within EntriesBackup on creation and to only restore an EntriesBackup-instance if the current context matches the originating context.

(Our workaround is to pin the junit-pioneer version to 1.4.0 for now.)

**Comment by beatngu13** [comment]
Thanks for reporting, will look into this.

**Comment by beatngu13** [comment]
So, I investigated that issue and can confirm that this indeed is a problem. But thanks to your excellent analysis (kudos!), it wasn't that hard to come up with a quick fix. As you already pointed out:

>A workaround could be to store something that identifies the originating context within EntriesBackup on creation and to only restore an EntriesBackup-instance if the current context matches the originating context.

I did this:

```java
private Object getStoreKey(ExtensionContext context) {
    return context.getDisplayName();
}
```

Determining the store key via the display name of the current test or container seems to do the trick.

Would you mind giving this a try? You can obtain [my fork's branch](https://github.com/beatngu13/junit-pioneer/tree/issue/480-nested-sysprop-envar) as a dependency e.g. via [JitPack](https://jitpack.io/#beatngu13/junit-pioneer/issue~480-nested-sysprop-envar-SNAPSHOT):

```xml
<repositories>
  <repository>
    <id>jitpack.io</id>
    <url>https://jitpack.io</url>
  </repository>
</repositories>

<!-- ... -->

<dependency>
  <groupId>com.github.beatngu13</groupId>
  <artifactId>junit-pioneer</artifactId>
  <version>issue~480-nested-sysprop-envar-SNAPSHOT</version>
</dependency>
```

**Comment by fknittel** [comment]
Thanks for your impressively quick response and fix! Looks good. (Still a bit unhappy, that JUnit doesn't provide a nicer/more obvious API for this scenario.)

I'll give your fork a spin, but I don't expect to uncover any surprises - your "failing test" looks sane.

**Comment by fknittel** [comment]
I can confirm that your fork/branch fixes my issue.

@beatngu13 Side-note: For some reason the [JitPack build](https://jitpack.io/#beatngu13/junit-pioneer/issue~480-nested-sysprop-envar-SNAPSHOT) contains [no artifacts](https://jitpack.io/com/github/beatngu13/junit-pioneer/issue~480-nested-sysprop-envar-v1.3.0-g0788c62-34/). While building locally, I had to disable signing to allow `publishToMavenLocal` to succeed, so maybe [some conditions](https://docs.gradle.org/current/userguide/signing_plugin.html#sec:conditional_signing) would be helpful?

**Comment by beatngu13** [comment]
Glad it worked out! 🥳 I just opened #481 and assume the fix will be shipped via v1.4.2.

Also thanks for pointing to the JitPack failure. We recently changed our build/publishing mechanism, which apparently leads to this problem. I will create a separate issue and PR for this.

**Comment by nipafx** [comment]
I just released 1.4.2 with the fix. I'm sorry it took 2.5 weeks (particularly after @beatngu13 created the fix in a day). 😔
